lib/ConfigurationHelpers.js

```diff
@@ -0,0 +1,37 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+'use strict'
+
+const { ConfigurationError } = require('./errors')
+
+function getAuth (nodes) {
+  nodes = nodes || []
+  nodes = Array.isArray(nodes) ? nodes : [nodes]
+
+  for (const node of nodes) {
+    const auth = getUsernameAndPassword(node)
+    if (auth.username) return auth
+  }
+
+  return null
+
+  function getUsernameAndPassword (node) {
+    let it = node && (node.url || node)
+    try {
+      it = it instanceof URL ? it : new URL((it && it.toString()))
+    } catch (e) {
+      throw new ConfigurationError(`invalid url in option "node${node.url ? '.url' : ''}": ${e.message}`)
+    }
+
+    return {
+      username: decodeURIComponent(it.username),
+      password: decodeURIComponent(it.password)
+    }
+  }
+}
+
+module.exports = {
+  getAuth
+}
```

Comment:
Let's rename this to `getCredentials()` and simplify a bit:

```js
// credentials.js

const { ConfigurationError } = require('./errors')

/**
 * Returns { username, password }, or throws ConfigurationError if node invalid
 * 
 * node can be:
 * A) 'http://abc:123@localhost:9200'
 * B) { url: new URL('http://abc:123@localhost:9200') }
 *
 * See: elasticsearch-js/test/unit/client.test.js
 */
function getCredentials (node) {
  try {
    const url = (node.url instanceof URL) ? node.url : new URL(node)
    return {
      username: decodeURIComponent(url.username),
      password: decodeURIComponent(url.password)
    }
  } catch (e) {
    throw new ConfigurationError(`invalid url in node "${node}": ${e.message}`)
  }
}

module.exports = { getCredentials }
```

Now, we can return credentials for the specific node only and move any other complexity (decision) outside, to `index.js`:

```js
// index.js

// ...
const node = Array.isArray(opts.nodes) ? opts.nodes[0] : opts.node
const credentials = getCrendentials(node)
// ...

// See also: elasticsearch-js/test/unit/client.test.js
```